transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,698 @@
+                                let ma = socketaddr_to_multiaddr(&socket_addr);
+                                debug!("New listen address: {}", ma);
+                                Poll::Ready(TransportEvent::NewAddress {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
+                            }
+                        }
+                        Ok(IfEvent::Down(inet)) => {
+                            let ip = inet.addr();
+                            if self.listen_addr.is_ipv4() == ip.is_ipv4()
+                                || self.listen_addr.is_ipv6() == ip.is_ipv6()
+                            {
+                                let socket_addr = SocketAddr::new(ip, self.listen_addr.port());
+                                let ma = socketaddr_to_multiaddr(&socket_addr);
+                                debug!("Expired listen address: {}", ma);
+                                Poll::Ready(TransportEvent::AddressExpired {
+                                    listener_id: self.listener_id,
+                                    listen_addr: ma,
+                                })
+                            } else {
+                                self.poll_if_addr(cx)
+                            }
+                        }
+                        Err(err) => {
+                            debug! {
+                                "Failure polling interfaces: {:?}.",
+                                err
+                            };
+                            Poll::Ready(TransportEvent::ListenerError {
+                                listener_id: self.listener_id,
+                                error: err.into(),
+                            })
+                        }
+                    }
+                } else {
+                    self.poll_if_addr(cx)
+                }
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl Stream for WebRTCListenStream {
+    type Item = TransportEvent<<WebRTCTransport as Transport>::ListenerUpgrade, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        if let Some(closed) = self.report_closed.as_mut() {
+            // Listener was closed.
+            // Report the transport event if there is one. On the next iteration, return
+            // `Poll::Ready(None)` to terminate the stream.
+            return Poll::Ready(closed.take());
+        }
+        if let Poll::Ready(event) = self.poll_if_addr(cx) {
+            return Poll::Ready(Some(event));
+        }
+
+        let udp_mux = self.udp_mux.clone();
+        let udp_mux_read_fut = self
+            .udp_mux_read_fut
+            .get_or_insert(Box::pin(async move { udp_mux.read_from_conn().await }));
```

Comment:
We are in control of the source code for `UdpMuxNewAddr` because we copied it in so we can just change this API to be better, i.e. `poll`-based, then we don't need this `Option<Future>` dance here.

Reply:
https://github.com/libp2p/rust-libp2p/pull/2622#discussion_r929684579

Reply:
I don't think this applies. If we store a plain UDP socket instead of a boxed `Conn`, we can rewrite `read_from_conn` to be poll-based instead of async which will then compose nicely here.